File: hw7/hw7.py
from typing import List


class HW7:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals = sorted(intervals, key = lambda  x : x[0])
        activeRooms = []
        maxRooms = 0
        for i in range(len(intervals)):
            j = len(activeRooms)
            while j > 0:
                if activeRooms[j - 1][1] <= intervals[i][0]:
                    activeRooms.pop(j - 1)
                j -= 1
            activeRooms.append(intervals[i])
            if len(activeRooms) > maxRooms:
                maxRooms = len(activeRooms)
        return maxRooms

    # O(n) solution
    def minMeetingRooms2(self, intervals: List[List[int]]) -> int:
        intervals = sorted(intervals, key=lambda x: x[1],reverse=True)
        startTimes = []
        endTimes = []
        for i in intervals:
            endTimes.append(i[1])
        intervals = sorted(intervals, key=lambda x: x[0],reverse=True)
        for i in intervals:
            startTimes.append(i[0])
        maxRooms = 0
        minRooms = 0
        while len(startTimes) > 0:
            time = 0
            if startTimes[-1] <= endTimes[-1]:
                time = startTimes[-1]
            else:
                time = endTimes[-1]
            while len(startTimes) > 0 and startTimes[-1] == time:
                startTimes.pop()
                minRooms += 1
            while len(endTimes) > 0 and endTimes[-1] == time:
                endTimes.pop()
                minRooms -= 1
            if minRooms > maxRooms:
                maxRooms = minRooms
        return maxRooms

hw7 = HW7()
interval = [[0,5],[5,10],[15,20]]
print(hw7.minMeetingRooms(interval))
print(hw7.minMeetingRooms2(interval))

File: hw7/hw7.java
package hw7;

import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Queue;

public class hw7 {
    public static int minMeetingRooms(int[][] intervals) {  // Your code here
        Arrays.sort(intervals, Comparator.comparingInt(o -> o[1]));
        Queue<Integer> endTimes = new LinkedList<>();
        for (int[] interval : intervals) {
            endTimes.add(interval[1]);
        }

        Arrays.sort(intervals, Comparator.comparingInt(o -> o[0]));
        Queue<Integer> startTimes = new LinkedList<>();
        for (int[] interval : intervals) {
            startTimes.add(interval[0]);
        }

        int maxRooms = 0;
        int minRooms = 0;

        while (!startTimes.isEmpty()){
            int time = 0;
            if (startTimes.peek() <= endTimes.peek()){
                time = startTimes.peek();
            }else{
                time = endTimes.peek();
            }

            while (!startTimes.isEmpty() && startTimes.peek() == time){
                startTimes.remove();
                minRooms++;
            }
            while (!endTimes.isEmpty() && endTimes.peek() == time){
                endTimes.remove();
                minRooms--;
            }

            if (minRooms > maxRooms){
                maxRooms = minRooms;
            }
        }

        return maxRooms;
    }

    public static void main(String[] args) {
        int[][] interval = {{0,30},{5,10},{15,20}};
        System.out.println(minMeetingRooms(interval));

        int[][] interval2 = {{0,1},{1,2},{2,3}};
        System.out.println(minMeetingRooms(interval2));
    }

}

